hathor/consensus/poa/poa_block_producer.py

```diff
@@ -0,0 +1,172 @@
+
+    def _start_producing(self) -> None:
+        """Start producing new blocks."""
+        if not self._manager.can_start_mining():
+            # We're syncing, so we'll try again later
+            self._log.warn('cannot start producing new blocks, node not synced')
+            return
+
+        self._log.info('started producing new blocks')
+        self._started_producing = True
+        self._start_producing_lc.stop()
+
+    def _schedule_block(self) -> None:
+        """Schedule propagation of a new block."""
+        previous_block = self._manager.tx_storage.get_best_block()
+        assert isinstance(previous_block, PoaBlock)
+        if not self._started_producing or previous_block == self._last_seen_best_block:
+            return
+
+        self._last_seen_best_block = previous_block
+        now = self._reactor.seconds()
+        expected_timestamp = self._expected_block_timestamp(previous_block)
+        propagation_delay = 0 if expected_timestamp < now else expected_timestamp - now
+
+        self._reactor.callLater(propagation_delay, self._produce_block, previous_block)
+        self._log.debug(
+            'scheduling block production',
+            previous_block=previous_block.hash_hex,
+            previous_block_height=previous_block.get_height(),
+            delay=propagation_delay,
+        )
+
+    def _produce_block(self, previous_block: PoaBlock) -> None:
+        """Create and propagate a new block."""
+        block_templates = self._manager.get_block_templates(parent_block_hash=previous_block.hash)
+        block = block_templates.generate_mining_block(self._manager.rng, poa_block=True)
+        assert isinstance(block, PoaBlock)
+        block.weight = poa.calculate_weight(self._poa_settings, block, self._signer_index)
+        self._poa_signer.sign_block(block)
+        block.update_hash()
+
+        self._manager.on_new_tx(block, propagate_to_peers=False, fails_silently=False)
+        if not block.get_metadata().voided_by:
+            self._manager.connections.send_tx_to_peers(block)
+
+        self._log.debug(
+            'produced new block',
+            block=block.hash_hex,
+            height=block.get_height(),
+            weight=block.weight,
+            parent=block.get_block_parent_hash().hex(),
+            voided=bool(block.get_metadata().voided_by),
+        )
+
+    def _expected_block_timestamp(self, previous_block: PoaBlock) -> int:
+        """Calculate the expected timestamp for a new block."""
+        height = previous_block.get_height() + 1
+        is_in_turn = poa.is_in_turn(settings=self._poa_settings, height=height, signer_index=self._signer_index)
+        timestamp = previous_block.timestamp + self._settings.AVG_TIME_BETWEEN_BLOCKS
+        if is_in_turn:
+            return timestamp
+
+        signer_count = len(self._poa_settings.signers)
+        assert signer_count >= 1
+        random_offset = self._manager.rng.choice(range(signer_count * _RANDOM_DELAY_MULTIPLIER)) + 1
```

Comment:
@msbrogli commented on https://github.com/HathorNetwork/hathor-core-side-dag/pull/11#discussion_r1618051818:

> This approach could potentially lead to orphan blocks. If the designated signer fails to propagate a new block promptly, other signers might simultaneously broadcast their blocks, resulting in conflicts.

Reply:
@msbrogli I changed it to use a round-robin increasing delay instead of a random delay in a separate PR, https://github.com/HathorNetwork/hathor-core/pull/1074

Reply:
Ok. Thanks!